File: app/src/main/java/com/kobrakid/retroachievements/fragment/SettingsFragment.java
package com.kobrakid.retroachievements.fragment;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.util.SparseArray;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.kobrakid.retroachievements.AppExecutors;
import com.kobrakid.retroachievements.Consts;
import com.kobrakid.retroachievements.MainActivity;
import com.kobrakid.retroachievements.R;
import com.kobrakid.retroachievements.RAAPICallback;
import com.kobrakid.retroachievements.RAAPIConnection;
import com.kobrakid.retroachievements.database.Console;
import com.kobrakid.retroachievements.database.RetroAchievementsDatabase;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class SettingsFragment extends Fragment implements RAAPICallback {

    private static final String TAG = SettingsFragment.class.getSimpleName();

    // Unused, but guarantees that the parent Activity implements OnFragmentInteractionListener
    @SuppressWarnings({"unused", "FieldCanBeLocal"})
    private OnFragmentInteractionListener listener;
    private RAAPIConnection apiConnection;
    private SharedPreferences sharedPref;
    private final StringBuilder consoleID = new StringBuilder();
    private final StringBuilder consoleName = new StringBuilder();

    private final SparseArray<Runnable> applicableSettings = new SparseArray<>();
    // Can be local, but used to index which settings have been modified
    @SuppressWarnings("FieldCanBeLocal")
    private final int logout_key = 0;
    private final int hide_consoles_key = 1;
    @SuppressWarnings("FieldCanBeLocal")
    private final int hide_games_key = 2;
    @SuppressWarnings("FieldCanBeLocal")
    private final int change_theme_key = 3;

    public SettingsFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        Objects.requireNonNull(getActivity()).setTitle("Settings");

        // Initialize preferences object
        sharedPref = Objects.requireNonNull(getContext()).getSharedPreferences(getString(R.string.shared_preferences_key), Context.MODE_PRIVATE);

        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_settings, container, false);

        // Set up views
        String theme = sharedPref.getString(getString(R.string.theme_setting), "");
        ((TextView) view.findViewById(R.id.settings_current_theme)).setText(getString(R.string.settings_current_theme, theme));
        ((TextView) view.findViewById(R.id.settings_current_user)).setText(MainActivity.ra_user == null ? getString(R.string.settings_no_current_user) : getString(R.string.settings_current_user, MainActivity.ra_user));

        ((Spinner) view.findViewById(R.id.settings_theme_dropdown)).setAdapter(new ArrayAdapter<String>(getContext(), android.R.layout.simple_spinner_dropdown_item, Consts.THEMES) {
            @Override
            public boolean isEnabled(int position) {
                return Consts.THEMES_ENABLE_ARRAY[position];
            }

            @Override
            public View getDropDownView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                TextView textView = (TextView) super.getDropDownView(position, convertView, parent);
                if (isEnabled(position)) {
                    TypedValue typedValue = new TypedValue();
                    getContext().getTheme().resolveAttribute(R.attr.colorPrimary, typedValue, true);
                    textView.setTextColor(getResources().getColor(typedValue.resourceId));
                } else {
                    textView.setTextColor(Color.GRAY);
                }
                return textView;
            }
        });
        ((Spinner) view.findViewById(R.id.settings_theme_dropdown)).setSelection(Arrays.asList(Consts.THEMES).indexOf(theme));
        ((Spinner) view.findViewById(R.id.settings_theme_dropdown)).setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int pos, long id) {
                if (pos > 0)
                    changeTheme(adapterView.getItemAtPosition(pos).toString());
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });

        ((CheckBox) view.findViewById(R.id.settings_hide_consoles)).setChecked(sharedPref.getBoolean(getString(R.string.empty_console_hide_setting), false));
        if (sharedPref.getBoolean(getString(R.string.empty_console_hide_setting), false))
            view.findViewById(R.id.settings_hide_consoles_warning).setVisibility(View.GONE);
        ((CheckBox) view.findViewById(R.id.settings_hide_games)).setChecked(sharedPref.getBoolean(getString(R.string.empty_game_hide_setting), false));
        ((CheckBox) view.findViewById(R.id.settings_hide_consoles)).setOnCheckedChangeListener((compoundButton, b) -> hideConsoles(view, b));
        ((CheckBox) view.findViewById(R.id.settings_hide_games)).setOnCheckedChangeListener((compoundButton, b) -> hideGames(view, b));

        apiConnection = ((MainActivity) getActivity()).apiConnection;

        return view;
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        try {
            listener = (OnFragmentInteractionListener) getActivity();
        } catch (ClassCastException e) {
            throw new ClassCastException(Objects.requireNonNull(getActivity()).toString()
                    + " must implement OnHeadlineSelectedListener");
        }
    }

    /* Settings-related Functions */

    private void changeTheme(@NonNull final String theme) {
        applicableSettings.remove(change_theme_key);
        applicableSettings.put(change_theme_key, () -> {
            Log.d(TAG, "Saving theme " + theme);
            sharedPref.edit().putString(getString(R.string.theme_setting), theme).apply();
        });
    }

    private void hideConsoles(View view, final boolean hide) {
        view.findViewById(R.id.settings_hide_consoles_warning).setVisibility(hide ? View.GONE : View.VISIBLE);
        if (applicableSettings.get(hide_consoles_key) != null) {
            applicableSettings.remove(hide_consoles_key);
        } else {
            final RAAPICallback callback = this;
            applicableSettings.put(hide_consoles_key, () -> {
                sharedPref.edit().putBoolean(getString(R.string.empty_console_hide_setting), hide).apply();
                if (hide) {
                    // Get all consoles and store their game counts
                    final RetroAchievementsDatabase db = RetroAchievementsDatabase.getInstance(getContext());
                    AppExecutors.getInstance().diskIO().execute(() -> {
                        db.consoleDao().clearTable();
                        Log.d(TAG, "Clearing table");
                    });
                    apiConnection.GetConsoleIDs(callback);
                } else {
                    Objects.requireNonNull(getActivity()).recreate();
                }
            });
        }
    }

    private void hideGames(View view, final boolean hide) {
        view.findViewById(R.id.settings_hide_games_warning).setVisibility(hide ? View.GONE : View.VISIBLE);
        if (applicableSettings.get(hide_games_key) != null) {
            applicableSettings.remove(hide_games_key);
        } else {
            applicableSettings.put(hide_games_key, () -> sharedPref.edit().putBoolean(getString(R.string.empty_game_hide_setting), hide).apply());
        }
    }

    public void logout() {
        ((TextView) Objects.requireNonNull(getActivity()).findViewById(R.id.settings_current_user)).setText(getString(R.string.settings_current_user, "none"));
        applicableSettings.put(logout_key, () -> sharedPref.edit().putString(getString(R.string.ra_user), null).apply());
    }

    public void applySettings() {
        Objects.requireNonNull(getActivity()).findViewById(R.id.settings_applying_fade).setVisibility(View.VISIBLE);
        getActivity().findViewById(R.id.settings_applying).setVisibility(View.VISIBLE);
        for (int key = 0; key < applicableSettings.size(); key++) {
            applicableSettings.valueAt(key).run();
        }
        if (applicableSettings.get(hide_consoles_key) == null) {
            // Recreate activity now if no db operations are running
            getActivity().recreate();
        }
    }

    @Override
    public void callback(int responseCode, final String response) {
        if (responseCode == RAAPIConnection.RESPONSE_ERROR)
            return;
        if (responseCode == RAAPIConnection.RESPONSE_GET_CONSOLE_IDS) {
            final RetroAchievementsDatabase db = RetroAchievementsDatabase.getInstance(getContext());
            final RAAPIConnection connection = apiConnection;
            final RAAPICallback callback = this;
            AppExecutors.getInstance().diskIO().execute(() -> {
                try {
                    JSONArray reader = new JSONArray(response);
                    for (int i = 0; i < reader.length(); i++) {
                        List<Console> consoles = db.consoleDao().getConsoleWithID(Integer.parseInt(reader.getJSONObject(i).getString("ID")));
                        if (consoles.size() == 0) {
                            consoleID.delete(0, consoleID.length());
                            consoleName.delete(0, consoleName.length());
                            consoleID.append(reader.getJSONObject(i).getString("ID"));
                            consoleName.append(reader.getJSONObject(i).getString("Name"));
                            connection.GetGameList(reader.getJSONObject(i).getString("ID"), callback);
                            return;
                        }
                    }
                    AppExecutors.getInstance().mainThread().execute(() -> Objects.requireNonNull(getActivity()).recreate());
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            });
        } else if (responseCode == RAAPIConnection.RESPONSE_GET_GAME_LIST) {
            try {
                final JSONArray reader = new JSONArray(response);
                final RetroAchievementsDatabase db = RetroAchievementsDatabase.getInstance(getContext());
                AppExecutors.getInstance().diskIO().execute(() -> {
                    db.consoleDao().insertConsole(new Console(Integer.parseInt(consoleID.toString()), consoleName.toString(), reader.length()));
                    Log.d(TAG, "Adding console " + consoleName.toString() + "(" + consoleID.toString() + "): " + reader.length() + " games");
                });
                // Recurse until all consoles are added to db
                apiConnection.GetConsoleIDs(this);
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }

    /* Inner Classes and Interfaces */

    public interface OnFragmentInteractionListener {
        void logout(@SuppressWarnings("unused") View view);

        void applySettings(@SuppressWarnings("unused") View view);
    }

}

File: app/src/main/java/com/kobrakid/retroachievements/fragment/SettingsFragment.kt
package com.kobrakid.retroachievements.fragment

import android.content.Context
import android.content.SharedPreferences
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.util.SparseArray
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import android.widget.AdapterView.OnItemSelectedListener
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.kobrakid.retroachievements.Consts
import com.kobrakid.retroachievements.R
import com.kobrakid.retroachievements.RetroAchievementsApi
import com.kobrakid.retroachievements.activity.MainActivity
import com.kobrakid.retroachievements.database.Console
import com.kobrakid.retroachievements.database.RetroAchievementsDatabase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONException

class SettingsFragment : Fragment(), View.OnClickListener {

    private var sharedPref: SharedPreferences? = null
    private val applicableSettings = SparseArray<Runnable?>()
    private var counter = 0

    // Can be local, but used to index which settings have been modified
    private val logoutKey = 0
    private val hideConsolesKey = 1
    private val hideGamesKey = 2
    private val changeThemeKey = 3

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        requireActivity().title = "Settings"
        return inflater.inflate(R.layout.fragment_settings, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        // Initialize preferences object
        sharedPref = context?.getSharedPreferences(getString(R.string.shared_preferences_key), Context.MODE_PRIVATE)

        // Set up views
        val theme = Consts.Theme.values().indexOfFirst { it.themeAttr == sharedPref?.getInt(getString(R.string.theme_setting), R.style.BlankTheme) }.coerceAtLeast(1)
        view.findViewById<TextView>(R.id.settings_current_theme).text = getString(R.string.settings_current_theme, Consts.Theme.values()[theme].themeName)
        view.findViewById<TextView>(R.id.settings_current_user).text = if (MainActivity.raUser == "") getString(R.string.settings_no_current_user) else getString(R.string.settings_current_user, MainActivity.raUser)
        view.findViewById<Spinner>(R.id.settings_theme_dropdown).adapter = object : ArrayAdapter<Consts.Theme>(requireContext(), android.R.layout.simple_spinner_dropdown_item, Consts.Theme.values()) {
            override fun isEnabled(position: Int): Boolean {
                return Consts.Theme.values()[position].enabled
            }

            override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
                val textView = super.getDropDownView(position, convertView, parent) as TextView
                if (isEnabled(position)) {
                    val typedValue = TypedValue()
                    context.theme.resolveAttribute(R.attr.colorPrimary, typedValue, true)
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
                        textView.setTextColor(context.resources.getColor(typedValue.resourceId, activity?.theme))
                    else
                        textView.setTextColor(ContextCompat.getColor(context, typedValue.resourceId))
                } else {
                    textView.setTextColor(Color.GRAY)
                }
                return textView
            }
        }
        view.findViewById<Spinner>(R.id.settings_theme_dropdown).setSelection(theme)
        view.findViewById<Spinner>(R.id.settings_theme_dropdown).onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View?, pos: Int, id: Long) {
                if (pos > 0) changeTheme(adapterView.getItemAtPosition(pos).toString())
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {}
        }
        view.findViewById<CheckBox>(R.id.settings_hide_consoles).isChecked = sharedPref?.getBoolean(getString(R.string.empty_console_hide_setting), false)!!
        if (sharedPref?.getBoolean(getString(R.string.empty_console_hide_setting), false) != false) view.findViewById<View>(R.id.settings_hide_consoles_warning).visibility = View.GONE
        view.findViewById<CheckBox>(R.id.settings_hide_games).isChecked = sharedPref?.getBoolean(getString(R.string.empty_game_hide_setting), false)!!
        view.findViewById<CheckBox>(R.id.settings_hide_consoles).setOnCheckedChangeListener { _: CompoundButton?, b: Boolean -> hideConsoles(view, b) }
        view.findViewById<CheckBox>(R.id.settings_hide_games).setOnCheckedChangeListener { _: CompoundButton?, b: Boolean -> hideGames(view, b) }

        view.findViewById<Button>(R.id.settings_logout).setOnClickListener(this)
        view.findViewById<Button>(R.id.settings_apply).setOnClickListener(this)
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.settings_logout -> logout()
            R.id.settings_apply -> applySettings()
        }
    }

    /* Settings-related Functions */
    private fun changeTheme(theme: String) {
        applicableSettings.remove(changeThemeKey)
        applicableSettings.put(changeThemeKey, Runnable {
            Log.d(TAG, "Saving theme $theme with theme attribute ${Consts.Theme.values()[Consts.Theme.values().indexOfFirst { it.themeName == theme }].themeAttr}")
            sharedPref?.edit()?.putInt(getString(R.string.theme_setting), Consts.Theme.values()[Consts.Theme.values().indexOfFirst { it.themeName == theme }].themeAttr)?.apply()
        })
    }

    private fun hideConsoles(view: View, hide: Boolean) {
        view.findViewById<View>(R.id.settings_hide_consoles_warning).visibility = if (hide) View.GONE else View.VISIBLE
        if (applicableSettings[hideConsolesKey] != null) {
            applicableSettings.remove(hideConsolesKey)
        } else {
            applicableSettings.put(hideConsolesKey, Runnable {
                sharedPref?.edit()?.putBoolean(getString(R.string.empty_console_hide_setting), hide)?.apply()
                if (hide) {
                    // Get all consoles and store their game counts
                    val db = context?.let { RetroAchievementsDatabase.getInstance(it) }
                    db?.let {
                        CoroutineScope(IO).launch {
                            db.consoleDao()?.clearTable()
                            Log.d(TAG, "Clearing console table")
                        }
                    }
                    val ctx = context?.applicationContext
                    CoroutineScope(IO).launch {
                        if (ctx != null)
                            RetroAchievementsApi.GetConsoleIDs(ctx) { removeConsoles(view, it) }
                    }
                } else {
                    activity?.recreate()
                }
            })
        }
    }

    private fun hideGames(view: View, hide: Boolean) {
        view.findViewById<View>(R.id.settings_hide_games_warning).visibility = if (hide) View.GONE else View.VISIBLE
        if (applicableSettings[hideGamesKey] != null) {
            applicableSettings.remove(hideGamesKey)
        } else {
            applicableSettings.put(hideGamesKey, Runnable { sharedPref?.edit()?.putBoolean(getString(R.string.empty_game_hide_setting), hide)?.apply() })
        }
    }

    private fun logout() {
        (activity?.findViewById<View>(R.id.settings_current_user) as TextView).text = getString(R.string.settings_current_user, "none")
        applicableSettings.put(logoutKey, Runnable { sharedPref?.edit()?.putString(getString(R.string.ra_user), "")?.apply() })
    }

    private fun applySettings() {
        activity?.findViewById<View>(R.id.settings_applying_fade)?.visibility = View.VISIBLE
        activity?.findViewById<View>(R.id.settings_applying)?.visibility = View.VISIBLE
        for (key in 0 until applicableSettings.size()) {
            applicableSettings.valueAt(key)?.run()
        }
        if (applicableSettings[hideConsolesKey] == null) {
            // Recreate activity now if no db operations are running
            activity?.recreate()
        }
    }

    private suspend fun removeConsoles(view: View, response: Pair<RetroAchievementsApi.RESPONSE, String>) {
        when (response.first) {
            RetroAchievementsApi.RESPONSE.ERROR -> {
                Log.w(TAG, response.second)
            }
            RetroAchievementsApi.RESPONSE.GET_CONSOLE_IDS -> {
                val db = context?.let { RetroAchievementsDatabase.getInstance(it) }
                try {
                    val reader = JSONArray(response.second)
                    counter = reader.length()
                    for (i in 0 until counter) {
                        val ctx = context?.applicationContext
                        if (db != null) {
                            CoroutineScope(IO).launch {
                                // Set each console to have 0 games
                                db.consoleDao()?.insertConsole(Console(reader.getJSONObject(i).getString("ID").toInt(), reader.getJSONObject(i).getString("Name"), 0))
                                if (ctx != null)
                                    RetroAchievementsApi.GetGameList(ctx, reader.getJSONObject(i).getString("ID")) {
                                        removeConsoles(view, it)
                                    }
                            }
                        }
                        Log.i(TAG, "calling api on $i")
                    }
                } catch (e: JSONException) {
                    Log.e(TAG, "Couldn't parse console IDs", e)
                }
            }
            RetroAchievementsApi.RESPONSE.GET_GAME_LIST -> {
                try {
                    val reader = JSONArray(response.second)
                    val db = context?.let { RetroAchievementsDatabase.getInstance(it) }
                    if (reader.length() > 0 && db != null) {
                        CoroutineScope(IO).launch {
                            // Only update a console if it has more than 0 games
                            val id = reader.getJSONObject(0).getString("ConsoleID").toInt()
                            val name = reader.getJSONObject(0).getString("ConsoleName")
                            db.consoleDao()?.updateConsole(Console(id, name, reader.length()))
                            Log.d(TAG, "Updating console $id ($name): ${reader.length()} games")
                        }
                    }
                } catch (e: JSONException) {
                    Log.e(TAG, "Couldn't parse game list", e)
                } finally {
                    counter--
                }
                if (counter == 0)
                    withContext(Main) { activity?.recreate() }
            }
            else -> {
                Log.v(TAG, "${response.first}: ${response.second}")
            }
        }
    }

    companion object {
        private val TAG = SettingsFragment::class.java.simpleName
    }
}